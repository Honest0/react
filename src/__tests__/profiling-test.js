// @flow

import type React from 'react';
import type ReactDOM from 'react-dom';
import typeof ReactTestRenderer from 'react-test-renderer';
import type Store from 'src/devtools/store';

describe('profiling', () => {
  let React: React;
  let ReactDOM: ReactDOM;
  let Scheduler;
  let SchedulerTracing;
  let TestRenderer: ReactTestRenderer;
  let store: Store;
  let utils;

  beforeEach(() => {
    utils = require('./utils');
    utils.beforeEachProfiling();

    store = global.store;
    store.collapseNodesByDefault = false;

    React = require('react');
    ReactDOM = require('react-dom');
    Scheduler = require('scheduler');
    SchedulerTracing = require('scheduler/tracing');
    TestRenderer = utils.requireTestRenderer();
  });

  describe('ProfilingSummary', () => {
    it('should be collected for each commit', async done => {
      const Parent = ({ count }) => {
        Scheduler.advanceTime(10);
        return new Array(count)
          .fill(true)
          .map((_, index) => <Child key={index} />);
      };
      const Child = () => {
        Scheduler.advanceTime(2);
        return null;
      };

      const container = document.createElement('div');

      utils.act(() => ReactDOM.render(<Parent count={2} />, container));
      utils.act(() => store.startProfiling());
      utils.act(() => ReactDOM.render(<Parent count={3} />, container));
      utils.act(() => ReactDOM.render(<Parent count={1} />, container));
      utils.act(() => ReactDOM.render(<Parent count={0} />, container));
      utils.act(() => store.stopProfiling());

      function Suspender({ rendererID, rootID }) {
        const profilingSummary = store.profilingCache.ProfilingSummary.read({
          rendererID,
          rootID,
        });
        expect(profilingSummary).toMatchSnapshot('ProfilingSummary');
        return null;
      }

      const rendererID = utils.getRendererID();
      const rootID = store.roots[0];

      await utils.actSuspense(() =>
        TestRenderer.create(
          <React.Suspense fallback={null}>
            <Suspender rendererID={rendererID} rootID={rootID} />
          </React.Suspense>
        )
      );

      done();
    });
  });

  describe('CommitDetails', () => {
    it('should be collected for each commit', async done => {
      const Parent = ({ count }) => {
        Scheduler.advanceTime(10);
        return new Array(count)
          .fill(true)
          .map((_, index) => <Child key={index} />);
      };
      const Child = () => {
        Scheduler.advanceTime(2);
        return null;
      };

      const container = document.createElement('div');

      utils.act(() => store.startProfiling());
      utils.act(() => ReactDOM.render(<Parent count={2} />, container));
      utils.act(() => ReactDOM.render(<Parent count={3} />, container));
      utils.act(() => ReactDOM.render(<Parent count={1} />, container));
      utils.act(() => ReactDOM.render(<Parent count={0} />, container));
      utils.act(() => store.stopProfiling());

      function Suspender({ commitIndex, rendererID, rootID }) {
        const commitDetails = store.profilingCache.CommitDetails.read({
          commitIndex,
          rendererID,
          rootID,
        });
        expect(commitDetails).toMatchSnapshot(
          `CommitDetails commitIndex: ${commitIndex}`
        );
        return null;
      }

      const rendererID = utils.getRendererID();
      const rootID = store.roots[0];

      for (let commitIndex = 0; commitIndex <= 3; commitIndex++) {
        await utils.actSuspense(() =>
          TestRenderer.create(
            <React.Suspense fallback={null}>
              <Suspender
                commitIndex={commitIndex}
                rendererID={rendererID}
                rootID={rootID}
              />
            </React.Suspense>
          )
        );
      }

      done();
    });
  });

  describe('FiberCommits', () => {
    it('should be collected for each rendered fiber', async done => {
      const Parent = ({ count }) => {
        Scheduler.advanceTime(10);
        return new Array(count)
          .fill(true)
          .map((_, index) => <Child key={index} />);
      };
      const Child = () => {
        Scheduler.advanceTime(2);
        return null;
      };

      const container = document.createElement('div');

      utils.act(() => store.startProfiling());
      utils.act(() => ReactDOM.render(<Parent count={1} />, container));
      utils.act(() => ReactDOM.render(<Parent count={2} />, container));
      utils.act(() => ReactDOM.render(<Parent count={3} />, container));
      utils.act(() => store.stopProfiling());

      function Suspender({ fiberID, rendererID, rootID }) {
        const fiberCommits = store.profilingCache.FiberCommits.read({
          fiberID,
          rendererID,
          rootID,
        });
        expect(fiberCommits).toMatchSnapshot(
          `FiberCommits: element ${fiberID}`
        );
        return null;
      }

      const rendererID = utils.getRendererID();
      const rootID = store.roots[0];

      for (let index = 0; index < store.numElements; index++) {
        await utils.actSuspense(() => {
          const fiberID = store.getElementIDAtIndex(index);
          if (fiberID == null) {
            throw Error(`Unexpected null ID for element at index ${index}`);
          }
          TestRenderer.create(
            <React.Suspense fallback={null}>
              <Suspender
                fiberID={fiberID}
                rendererID={rendererID}
                rootID={rootID}
              />
            </React.Suspense>
          );
        });
      }

      done();
    });
  });

  describe('Interactions', () => {
    it('should be collected for every traced interaction', async done => {
      const Parent = ({ count }) => {
        Scheduler.advanceTime(10);
        return new Array(count)
          .fill(true)
          .map((_, index) => <Child key={index} />);
      };
      const Child = () => {
        Scheduler.advanceTime(2);
        return null;
      };

      const container = document.createElement('div');

      utils.act(() => store.startProfiling());
      utils.act(() =>
        SchedulerTracing.unstable_trace(
          'mount: one child',
          Scheduler.unstable_now(),
          () => ReactDOM.render(<Parent count={1} />, container)
        )
      );
      utils.act(() =>
        SchedulerTracing.unstable_trace(
          'update: two children',
          Scheduler.unstable_now(),
          () => ReactDOM.render(<Parent count={2} />, container)
        )
      );
      utils.act(() => store.stopProfiling());

      function Suspender({ rendererID, rootID }) {
        const interactions = store.profilingCache.Interactions.read({
          rendererID,
          rootID,
        });
        expect(interactions).toMatchSnapshot('Interactions');
        return null;
      }

      const rendererID = utils.getRendererID();
      const rootID = store.roots[0];

      await utils.actSuspense(() =>
        TestRenderer.create(
          <React.Suspense fallback={null}>
            <Suspender rendererID={rendererID} rootID={rootID} />
          </React.Suspense>
        )
      );

      done();
    });
  });

  it('should remove profiling data when roots are unmounted', async () => {
    const Parent = ({ count }) =>
      new Array(count).fill(true).map((_, index) => <Child key={index} />);
    const Child = () => <div>Hi!</div>;

    const containerA = document.createElement('div');
    const containerB = document.createElement('div');

    utils.act(() => {
      ReactDOM.render(<Parent key="A" count={3} />, containerA);
      ReactDOM.render(<Parent key="B" count={2} />, containerB);
    });

    utils.act(() => store.startProfiling());

    utils.act(() => {
      ReactDOM.render(<Parent key="A" count={4} />, containerA);
      ReactDOM.render(<Parent key="B" count={1} />, containerB);
    });

    utils.act(() => ReactDOM.unmountComponentAtNode(containerB));

    utils.act(() => ReactDOM.unmountComponentAtNode(containerA));

    utils.act(() => store.stopProfiling());

    // Assert all maps are empty
    store.assertExpectedRootMapSizes();
  });
});
