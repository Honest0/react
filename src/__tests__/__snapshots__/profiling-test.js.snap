// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`profiling CommitDetails should be collected for each commit: CommitDetails commitIndex: 0 1`] = `
Object {
  "actualDurations": Map {
    1 => 14,
    2 => 14,
    3 => 2,
    4 => 2,
  },
  "commitIndex": 0,
  "interactions": Array [],
  "rootID": 1,
}
`;

exports[`profiling CommitDetails should be collected for each commit: CommitDetails commitIndex: 1 1`] = `
Object {
  "actualDurations": Map {
    3 => 2,
    4 => 2,
    5 => 2,
    2 => 16,
    1 => 16,
  },
  "commitIndex": 1,
  "interactions": Array [],
  "rootID": 1,
}
`;

exports[`profiling CommitDetails should be collected for each commit: CommitDetails commitIndex: 2 1`] = `
Object {
  "actualDurations": Map {
    3 => 2,
    2 => 12,
    1 => 12,
  },
  "commitIndex": 2,
  "interactions": Array [],
  "rootID": 1,
}
`;

exports[`profiling CommitDetails should be collected for each commit: CommitDetails commitIndex: 3 1`] = `
Object {
  "actualDurations": Map {
    2 => 10,
    1 => 10,
  },
  "commitIndex": 3,
  "interactions": Array [],
  "rootID": 1,
}
`;

exports[`profiling FiberCommits should be collected for each rendered fiber: FiberCommits: element 2 1`] = `
Object {
  "commitDurations": Array [
    0,
    12,
    1,
    14,
    2,
    16,
  ],
  "fiberID": 2,
  "rootID": 1,
}
`;

exports[`profiling FiberCommits should be collected for each rendered fiber: FiberCommits: element 3 1`] = `
Object {
  "commitDurations": Array [
    0,
    2,
    1,
    2,
    2,
    2,
  ],
  "fiberID": 3,
  "rootID": 1,
}
`;

exports[`profiling FiberCommits should be collected for each rendered fiber: FiberCommits: element 4 1`] = `
Object {
  "commitDurations": Array [
    1,
    2,
    2,
    2,
  ],
  "fiberID": 4,
  "rootID": 1,
}
`;

exports[`profiling FiberCommits should be collected for each rendered fiber: FiberCommits: element 5 1`] = `
Object {
  "commitDurations": Array [
    2,
    2,
  ],
  "fiberID": 5,
  "rootID": 1,
}
`;

exports[`profiling Interactions should be collected for every traced interaction: Interactions 1`] = `
Array [
  Object {
    "__count": 0,
    "commits": Array [
      1,
    ],
    "id": 0,
    "name": "one child",
    "timestamp": 10,
  },
  Object {
    "__count": 0,
    "commits": Array [
      2,
    ],
    "id": 1,
    "name": "two children",
    "timestamp": 22,
  },
]
`;

exports[`profiling ProfilingSummary should be collected for each commit: ProfilingSummary 1`] = `
Object {
  "commitDurations": Array [
    16,
    12,
    10,
  ],
  "commitTimes": Array [
    16,
    28,
    38,
  ],
  "initialTreeBaseDurations": Map {
    1 => 14,
    2 => 14,
    3 => 2,
    4 => 2,
  },
  "interactionCount": 0,
  "rootID": 1,
}
`;
